### PR TITLE
feat(batcher): add batching for CloudProvider Create and Delete

### LINKED ISSUE #87: CloudProvider Create and Delete should not block
### Description

**What problem are you trying to solve?**
Our karpenter cluster api provider currently blocks concurrent instance creations because of a global mutex in our [Create](https://github.com/kubernetes-sigs/karpenter-provider-cluster-api/blob/f0e21b7bffb4e804a4d4e2269fb08c8abfb6ae56/pkg/cloudprovider/cloudprovider.go#L89) function. This mutex is used in order to protect the `replicas` field on the `MachineDeployment`, which Karpenter CAPI updates for machine provisioning and de-provisioning.

Currently, our karpenter-capi-provider flow is that when we need to create provider-specific compute based on a pending pod, we:
1. Find all instance types (in karpenter-capi, these are `MachineDeployments`) that match the `NodeClaim` requirements.
2. If it finds a compatible instance type, we will provision an instance of it.
    - in karpenter capi, this means that we find a compatible `MachineDeployment`, and we scale the replicas of the deployment up 1
3. CAPI will then create the Machine for us asynchronously
4. Karpenter CAPI then blocks the `Create` path and polls for this `Machine` to be created, [before labelling it with a special label denoting it as claimed](https://github.com/kubernetes-sigs/karpenter-provider-cluster-api/blob/f0e21b7bffb4e804a4d4e2269fb08c8abfb6ae56/pkg/cloudprovider/cloudprovider.go#L373-L381)

We can see that if we have a large burst of pods that all create more machines from the same `MachineDeployment`, for example 10 pods creating 10 `Machine`s, this quickly becomes a problem where the CAPI provider will be constantly blocking the `Create` path and the last unlucky 10th pod may wait potentially 10 minutes (poll timeout is 60 seconds) for a Machine to be provisioned because of the sync/async mechanism. Even if 10 pods each brought up its own `Machine` from 10 distinct `MachineDeployment`s, we still suffer the waiting problem because the mutex is global.

Additionally the same mutex is implemented in the `Delete` path as well, so both `Create` and `Delete` paths will contend with each other as well.

**Solution?**
I propose that we use a Batcher to solve this problem. In essence, instead of globally locking the replicas field, and synchronously creating and deleting, we can use the idea of a Batcher which will collect inputs (in this case, calls to update the `MachineDeployment.spec.replicas`) within a "batch window", and eventually execute a single API call which aggregates all of the collected inputs. We can combine inputs into the same batch by using a batchKey based on the `MachineDeployment` name + namespace, since that is the contentious resource here. Concretely, this means if we have 10 pods that creates 10 nodes from the same `MachineDeployment` all in the span of a short window (e.g., 1 second), we can aggregate those calls into 1 simple kube-api call which updates the replicas field, instead of 10 of these calls. Since batches are keyed by `MachineDeployment`, requests targeting different `MachineDeployment`s would naturally execute as independent concurrent batches, eliminating the global lock entirely. The batcher would run as a infinite loop goroutine, and constantly be watching for inputs, and we provide inputs from the `cloudprovider.Create` implementation.

However we still have to account for polling, as in when we wait for CAPI to actually create the machines. To solve this, instead of each `Create` individually polling for an unclaimed machine, the batcher should poll for `batch` number of unclaimed machines. If only 6 machines have been claimed and labelled out of the 10 by the end of the timeout, we can rollback the machine deployment replicas back to 6, report the 4 launched `NodeClaims` as failures, and Karpenter core will delete the `NodeClaims` and try again.

This concurrent batcher idea architecture/idea exists already in the [karpenter-provider-aws](https://github.com/aws/karpenter-provider-aws/blob/5e978c594ecefa75bbae75ac3b25ed71294901c6/pkg/batcher/batcher.go#L60), and I believe it makes sense here. I also like re-using designs across the Karpenter ecosystem :-) As an example, one batcher implementation exist where Karpenter AWS tries to [minimize the number of AWS CreateFleet API calls ](https://github.com/aws/karpenter-provider-aws/blob/5e978c594ecefa75bbae75ac3b25ed71294901c6/pkg/batcher/createfleet.go#L36)on the same launch configuration (same launch template, instance type, capacity type, etc.) by batching them.

**Other**:
Concurrency also causes weird stuff with when doing complex provisioning. We can see these failures in this e2e run in my WIP PR here: https://github.com/kubernetes-sigs/karpenter-provider-cluster-api/actions/runs/25143532000/job/73698397569

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by maxcao13**
Fixes #87 

**Description**
Replace the mutex-serialized Create/Delete with a time-windowed batching architecture. Concurrent requests targeting the same MachineDeployment are coalesced into a single replica read-modify-write cycle, reducing optimistic concurrency conflicts that caused replica leaks and reducing lock scope from the entire create/poll flow to just the replica update.

Create batches count unclaimed Machines and increment replicas only by the deficit, so retries never over-provision. Delete batches annotate Machines in parallel and decrement replicas in one write with rollback of the annotations on failure (we don't rollback replica changes).

This PR adds:
- `pkg/batcher/` — generic `Batcher[T, U]` (adapted from karpenter-provider-aws), plus CAPI-specific `CreateBatcher` and `DeleteBatcher`
- `MDLockManager` — per-MachineDeployment mutex that serializes replica updates between create and delete batches
- `providers.MachineAnnotation` + `ParseMachineAnnotation` — moved the machine annotation constant out of cloudprovider into a shared location
- unit tests testing core functionality and corner cases of the each batcher individually as well as the create + delete batching interactions together

Other notes:
- The create executor's "count unclaimed first" strategy means there is no explicit replica rollback on partial failure — retries from the lifecycle controller will see the existing unclaimed Machines and not increment again. The algorithm is documented on `execCreateBatch`.
- `bindMachineToNodeClaim` retries the Machine label Update up to 3 times in the case of optimistic concurrency failures (stale resourceVersion).
- The poll step (`pollForNUnclaimedMachines`) runs unlocked with a 30s timeout. A concurrent delete batch can cause the poll to find fewer Machines than expected, but unfulfilled requests will simply error out and retry naturally.

Assisted with Cursor agent with Claude 4.6 opus model.

**How was this change tested?**

Regressions performance E2E tests and new batcher unit tests included in this PR. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .github/workflows/e2e.yaml [hunk 1]
```diff
@@ -28,8 +28,30 @@ jobs:
         ./test/hack/e2e.sh
       env:
         KIND_CLUSTER_IMAGE: kindest/node:v1.32.8@sha256:abd489f042d2b644e2d033f5c2d900bc707798d075e8186cb65e3f1367a9d5a1
+    - name: install controller
```
Comment (maxcao13) [inline .github/workflows/e2e.yaml]: I'll separate e2es out of this PR when its ready to merge. This is just to show that these changes work and allow the new e2es to pass: https://github.com/kubernetes-sigs/karpenter-provider-cluster-api/pull/86

**Comment by elmiko** [comment]
very cool! it will probably take me a few days to review this.

**Comment by maxcao13** [comment]
removed the e2e commits, they will come in a followup

File: pkg/batcher/create.go [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+		machines := pollForNUnclaimedMachines(ctx, machineProvider, mdName, mdNS, n, 30*time.Second)
+
+		// 3) Bind each Machine to a NodeClaim in parallel.
+		var wg sync.WaitGroup
+		for i := 0; i < len(machines) && i < n; i++ {
+			wg.Add(1)
+			go func(idx int) {
+				defer wg.Done()
+				results[idx] = bindMachineToNodeClaim(ctx, kubeClient, machineProvider, md, machines[idx], inputs[idx].NodeClaimName)
+			}(i)
```
Comment (elmiko) [inline pkg/batcher/create.go]: in reading up on `sync.WaitGroup` i came across the guidance that `WaitGroup.Go` is preferred to the `Add/Done` pattern. i wonder if might use that pattern to make this a little simpler?

all together it might look like:

```suggestion
			bindFn := func() {
			    results[i] = bindMachineToNodeClaim(ctx, kubeClient, machineProvider, md, machines[i], inputs[i].NodeClaimName)
			}
			wg.Go(bindFn)
```

File: pkg/batcher/delete.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+				if err != nil {
+					results[idx] = Result[DeleteOutput]{Err: fmt.Errorf("unable to get Machine %q: %w", machineName, err)}
+					return
+				}
+				if err := machineProvider.AddDeleteAnnotation(ctx, fresh); err != nil {
+					results[idx] = Result[DeleteOutput]{Err: fmt.Errorf("unable to annotate Machine %q for deletion: %w", machineName, err)}
+					return
+				}
+				annotated[idx] = true
+			}(i, input.MachineName, input.MachineNamespace)
```
Comment (elmiko) [inline pkg/batcher/delete.go]: similar comment here as in create.go, but the diff would look like:


```suggestion
			bindFn := func() {
				fresh, err := machineProvider.Get(ctx, input.MachineName, input.MachineNamespace)
				if err != nil {
					results[i] = Result[DeleteOutput]{Err: fmt.Errorf("unable to get Machine %q: %w", input.MachineName, err)}
					return
				}
				if err := machineProvider.AddDeleteAnnotation(ctx, fresh); err != nil {
					results[i] = Result[DeleteOutput]{Err: fmt.Errorf("unable to annotate Machine %q for deletion: %w", input.MachineName, err)}
					return
				}
				annotated[i] = true
			}
            wg.Go(bindFn)
```

File: pkg/providers/providers.go [hunk 4]
```diff
@@ -16,7 +16,33 @@ limitations under the License.
...
 	// NodePoolMemberLabel is a label used to identify resources that are used by karpenter
 	NodePoolMemberLabel = "node.cluster.x-k8s.io/karpenter-member"
+
+	// MachineAnnotation is the annotation on a NodeClaim that references
+	// the CAPI Machine bound to it, in "namespace/name" format.
+	MachineAnnotation = "cluster.x-k8s.io/machine"
 )
+
+// ParseMachineAnnotation splits a "namespace/name" annotation value into its components.
+func ParseMachineAnnotation(annotationValue string) (string, string, error) {
```
Comment (elmiko) [inline pkg/providers/providers.go:34]: do we have a unit test for this? (it's fairly small, but there is some logic in here)

**Review by elmiko — COMMENTED** [review]
this is mostly making sense to me, and i think it's an excellent addition. i have a couple minor suggestions and question.

File: pkg/batcher/create.go (hunk 2 above)
Comment (maxcao13) [reply pkg/batcher/create.go]: Oh neat :-)
I think this was introduced in go1.25 though, and this project is currently at 1.24: https://go.dev/doc/go1.25#syncpkgsync

Should I apply this in a followup when we bump go?

Comment (elmiko) [reply pkg/batcher/create.go]: yeah, why don't we just put a `//TODO` for now.
